File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.py
def end(index,n):
    #print(index, n)
    return [(index%n)+1, (index//n)+1]
def solution(n, words):
    # 현재 단어의 첫 글자가 이전 단어의 마지막 글자와 같은지 확인
    # 이전에 나왔던 단어인지 확인 -> set에 나온 단어 저장
    # 인덱스를 n으로 나누고 + 1을 한 것이 번호
    # 차례는 인덱스를 n으로 나눈 몫 + 1
    previous_c = words[0][-1]
    previous_words = set([])
    previous_words.add(words[0])
    for i in range(1,len(words)):
        if words[i][0] != previous_c or words[i] in previous_words:
            return end(i,n)
        previous_c = words[i][-1]
        previous_words.add(words[i])
    return [0,0]

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*;
class Solution {
    public int[] solution(int n, String[] words) {
        /*
        끝말잇기 규칙
        탈락한 사람의 번호/회차
        
        1. 이전 단어의 마지막 단어와 일치여부 확인
        2. 중복 여부 확인
        
        words를 순회하면서 순서와 회차 카운팅
        
        끝까지 지켰으면 [0,0]
        */
        Set<String> used = new HashSet<>();
        
        int order = 0;
        int round = 0;
        char last = '\0';
        for(String word: words){
            order %= n;
            // 순서가 돌아옴
            if(order == 0){
                round += 1;
            }
            //System.out.println(order +" " + round);
            char first = word.charAt(0);
            // 규칙 위반
            if(!(last == '\0' || last == first) || used.contains(word)){
                return new int[]{order+1, round};
            }
            last = word.charAt(word.length()-1);
            order++;
            used.add(word);
        }
        return new int[]{0,0};
    }
}